### ISSUE TITLE
Strategy interface proposal

### ISSUE THREAD

**Issue by AntonioCarta** [ContinualAI/avalanche#38]
I am starting to play around with the Strategy class and I would like to propose some changes:

- We need a method to change the loss function and add regularizers. Currently, if you want to define a regularization-based method you must necessarily redefine the `train` method with the entire training loop. This causes a lot of code duplication. I would like to have a `compute_loss` method that gets called inside `train`.
- The callbacks (`after_train`, `before_train`, ...) are all implemented as abstract methods. This means that each new strategy must define the method itself. Most of the strategies will implement these as empty methods. Can't we give a default empty implementation?
- The `multi_head` is not used. What is it doing?

Finally, I think that we should try to separate logging code (tensorboard, print statements) from training code. The EvalProtocol should be the only one doing the logging. However, this is less urgent right now.

I can do the changes, but first I wanted to discuss them with you.
Notice that already existing code is not affected by these changes.
